### PR TITLE
[DSCP-467] Auto swagger

### PR DESCRIPTION
**Pull request by Martoon-00**
### Description

**Do not merge: depends not on master of `servant-util`**

Here I add automatic generation of swagger documentation for Educator and Student APIs which is pretty close to one we currently have.
Also, an executable which dumps swagger specifications is added. In DSCP-473 I will also add swagger UI serving this documentation next to the main endpoints.

Compare:
* Student API: [manual doc](https://editor.swagger.io?url=https://raw.githubusercontent.com/DisciplinaOU/disciplina/master/specs/disciplina/educator/api/student.yaml) vs [automatically generated doc](https://editor.swagger.io?url=https://gist.githubusercontent.com/Martoon-00/4dd2e94a0658216cf9be248f53715312/raw/7e1aefba51da2b1bc6ebcbf7d9109472200a999b/student.json) 
* Educator API: [manual doc](https://editor.swagger.io?url=https://raw.githubusercontent.com/DisciplinaOU/disciplina/master/specs/disciplina/educator/api/educator.yaml) vs [automatically generated doc](https://editor.swagger.io?url=https://gist.githubusercontent.com/Martoon-00/4dd2e94a0658216cf9be248f53715312/raw/7e1aefba51da2b1bc6ebcbf7d9109472200a999b/educator.json) 
**Do not open links too quick. Auto docs should be in json format, manual ones in yaml.**

#### Development side effects
I needed to do some changes in the surrounding code, but they are either non-invasive or make sense for other reasons (as far as I believe). Examples:
* Added phantom parameter to `DocumentType` - it helps to distinguish the exact purpose of this thing when appearing as part of endpoint parameter but also helps you understand what's going on in code faster.
* Wrapped `LByteString` of `GET certificates` endpoint into a newtype. Swagger _forbids_ ByteStrings as endpoint parameters/response (i.e. compile error is raised). But generally, logging is made saner in result IMO (no more `instance Buildable LByteString` anymore). 

### YT issue

https://issues.serokell.io/issue/DSCP-467

### Checklist

Hint: a perfect PR has all the checkmarks set.

Possible related changes (conditional):
- [x] I added tests if required, in case they are:
  - Covering new introduced functionality (feature).
  - Regression tests preventing the bug from silently reappearing again (bugfix).
- [x] I have checked the [sample config](../tree/master/docs/config-full-sample.yaml) and [launch scripts](../tree/master/scripts/launch) and updated them if it was necessary (especially if executables or CLIs were changed).
- [x] I have checked READMEs and changed them accordingly if it's necessary (the main [README.md](../tree/master/README.md) as well as subproject READMEs for all related executables).
- [x] If a public API structure or/and data serialization was changed, I made sure that these changes are reflected in the:
  - [Swagger](../tree/master/specs/disciplina) API specs.
  - Any [related documentation](../tree/master/docs/api-types.md).
- [x] If any public documentation was written, I have spellchecked it.

Stylistic (obligatory):
- [x] My commit history is clean, descriptive and do not contain merge or revert commits or commits like "WIP".
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).
- [x] My code is documented (haddock) and these docs are decent (full enough and spellchecked).

### CODE REVIEW

File: educator/src/Dscp/Educator/Web/Auth.hs [hunk 1]
```diff
@@ -336,3 +342,61 @@ instance FromJSON (NoAuthData s) => FromJSON (NoAuthContext s) where
...
+                }
+            , S._securitySchemeDescription = Just educatorAuthDocDesc
+            }
+        response401Name = "Unauthorized"
+        response401 = mempty
+            & S.description .~ "Unauthorized"
+            & S.headers . at "WWW-Authenticate" ?~
+                (mempty & S.type_ .~ S.SwaggerString)
+
+-- TODO: this looks :cry:
```
Comment (pasqu4le) [inline educator/src/Dscp/Educator/Web/Auth.hs]: :rofl:

File: educator/src/Dscp/Educator/Web/Student/API.hs [hunk 2]
```diff
@@ -98,46 +110,62 @@ type GetAssignment
...
+    :> ExceptionalResponses StudentAPIError
+       '[ 403 #: "Cannot delete a graded submission"
+        , 404 #: "Submission with given hash not found"
+        ]
     :> Verb 'DELETE 200 '[DSON] ()
 
 ---------------------------------------------------------------------------
 -- Proofs
----------------------------------------------------------------------------
+--------------------------------------------------------------------------
```
Comment (pasqu4le) [inline educator/src/Dscp/Educator/Web/Student/API.hs]: Unacceptable.

File: educator/src/Dscp/Educator/Web/Types.hs [hunk 3]
```diff
@@ -195,3 +197,36 @@ deriving instance FromHttpApiData IsGraded
...
 deriving instance ToHttpApiData IsFinal
 deriving instance ToHttpApiData IsGraded
+
+---------------------------------------------------------------------------
+-- Swagger instances
+---------------------------------------------------------------------------
+
+type instance ParamDescription IsEnrolled =
+    "`True` when the student is currently enrolled to the course, \
+    \`False` when the course is yet only available for the student."
```
[the remark this answers is not in the document]
Comment (Martoon-00) [reply educator/src/Dscp/Educator/Web/Types.hs]: Good point, this is the doc for the frontend after all.

File: educator/src/Dscp/Educator/Web/Student/API.hs [hunk 4]
```diff
@@ -74,19 +80,25 @@ type GetCourse
...
+                 \look it up)"
+        ]
     :> Verb 'GET 200 '[DSON] [AssignmentStudentInfo]
 
 type GetAssignment
     = "assignments" :> Capture "assignment" (Hash Core.Assignment)
+    :> Tag "Assignments"
     :> Summary "Get info about an assignment"
     :> Description "Gets an assignment info by given submission hash. Returns \
                    \404 if a student tries to get an assignment which is not assigned to them."
```
[the remark this answers is not in the document]
Comment (Martoon-00) [reply educator/src/Dscp/Educator/Web/Student/API.hs]: Nice catch :+1:

[the remark this answers is not in the document]
Comment (pasqu4le) [reply educator/src/Dscp/Educator/Web/Student/API.hs]: I see you have added this but removed `Description` entirely, why?
Wouldn't it be better to have a shorter one like you did with `GetSubmission`?

[the remark this answers is not in the document]
Comment (Martoon-00) [reply educator/src/Dscp/Educator/Web/Student/API.hs]: Oh right, dunno what I'm doing

**Pull request merged**